scripts: gen_kobject_list: linker section for _thread_idx_map

dcpleung opened the pull request:
In some circumstances, _thread_idx_map[] is all zero and the linker
decides to put it into BSS instead of DATA section. This results in
kernel objects being pushed away so the hash table is no longer
valid. This forces _thread_idx_map to be in the data section inside
the intermediate object file so it will be placed in the data
section in final binary.

Fixes #43618

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: gen_kobject_list: linker section for _thread_idx_map

In some circumstances, _thread_idx_map[] is all zero and the linker
decides to put it into BSS instead of DATA section. This results in
kernel objects being pushed away so the hash table is no longer
valid. This forces _thread_idx_map to be in the data section inside
the intermediate object file so it will be placed in the data
section in final binary.

Fixes #43618

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Compliance check failed on line not touched by this PR.

andyross reviewed (dismissed):
Nice.  This explains nicely why the problem only shows up when the array is full -- the default value set elsewhere in the script is 0xff.

Note that some variant of this issue still likely exists.  The _thread_idx_map[] seen at runtime is this one generated here, but the only seen **by** the generator happens in the linker script here:  https://github.com/zephyrproject-rtos/zephyr/blob/main/include/linker/kobject-data.ld#L39

And that one has zero size, which isn't going to match the runtime one of CONFIG_MAX_THREAD_BYTES.  We seem to be saved currently by the much coarser alignment to KOBJECT_DATA_ALIGN which hides the mismatch.

We surely want to be generating this in a more consistent way somehow, but that can happen later.

nashif commented:
> Compliance check failed on line not touched by this PR.

can you please still fix it?

dcpleung committed:
scripts: gen_kobject_list: fix pylint complain on using keys()

This fixes a pylint complain on using .keys() while looping through
dictionary.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

andyross reviewed (commented):
Still +1
